mlfinlab/online_portfolio_selection/mean_reversion/online_moving_average_reversion.py

```diff
@@ -0,0 +1,111 @@
+
+class OnlineMovingAverageReversion(OLPS):
+    """
+    This class implements the Online Moving Average Reversion strategy. It is reproduced with
+    modification from the following paper:
+    `Li, Bin & Hoi, Steven. (2012). On-Line Portfolio
+    Selection with Moving Average Reversion. Proceedings of the 29th International Conference on
+    Machine Learning, ICML 2012. 1. <https://arxiv.org/pdf/1206.4626.pdf>`_
+
+    Online Moving Average Reversion reverts to the SMA or EMA of the underlying assets based on
+    the given threshold.
+    """
+
+    def __init__(self, reversion_method, epsilon, window=None, alpha=None):
+        """
+        Initializes Online Moving Average Reversion with the given reversion method, epsilon,
+        window, and alpha.
+
+        :param reversion_method: (int) 1 for SMA, 2 for EWA.
+        :param epsilon: (float) Reversion threshold. Must be greater or equal to 1.
+        :param window: (int) Number of windows to calculate Simple Moving Average.
+        :param alpha: (float) Ratio between 0 and 1 for Exponentially Weighted Average.
+        """
+        self.epsilon = epsilon
+        self.window = window
+        self.alpha = alpha
+        self.reversion_method = reversion_method
+        self.moving_average_reversion = None  # (np.array) Calculated moving average.
+        super().__init__()
+
+    def _initialize(self, asset_prices, weights, resample_by):
+        """
+        Initializes the important variables for the object.
+
+        :param asset_prices: (pd.DataFrame) Historical asset prices.
+        :param weights: (list/np.array/pd.Dataframe) Initial weights set by the user.
+        :param resample_by: (str) Specifies how to resample the prices.
+        """
+        super(OnlineMovingAverageReversion, self)._initialize(asset_prices, weights, resample_by)
+
+        # Check that reversion method is either 1 or 2.
+        if self.reversion_method not in [1, 2]:
+            raise ValueError("Reversion method must be either 1 or 2.")
+
+        # Check that epsilon values are correct.
+        if self.epsilon < 1:
+            raise ValueError("Epsilon values must be greater or equal to 1.")
+
+        # Check that window is at least 1.
+        if self.reversion_method == 1 and self.window < 1:
+            raise ValueError("Window must be at least 1.")
+
+        # Chat that alpha is between 0 and 1 for method 2.
+        if self.reversion_method == 2 and (self.alpha < 0 or self.alpha > 1):
+            raise ValueError("Alpha must be between 0 and 1.")
+
+        # Calculate moving average reversion.
+        self.moving_average_reversion = self._calculate_rolling_moving_average()
+
+    def _update_weight(self, time):
+        """
+        Predicts the next time's portfolio weight.
+
+        :param time: (int) Current time period.
+        :return new_weights: (np.array) Predicted weights.
+        """
+        # Return predetermined weights for time periods with no significant data.
```

Comment:
This code is a little bit clumped together. Add whitespaces to separate sections much like we use paragraphs in language. this will improve readability.

Reply:
Also breaking this up into smaller functions will help but let's do that going forward.